### ISSUE TITLE
Using pkcs11-tool with Infineon SLB9670 TPM fails object enumeration, key generation, self-test

### ISSUE THREAD

**Issue by pfyvie** [tpm2-software/tpm2-pkcs11#264]
Performing operations through pkcs11-tool on a hardware TPM (Infineon SLB9670) fails with various errors starting with `CKR_` (see terminal output below). I was following https://github.com/tpm2-software/tpm2-pkcs11/blob/master/docs/PKCS11_TOOL.md when I encountered this.

The main error I'm seeing is `CKR_ATTRIBUTE_TYPE_INVALID`, although the self-test also throws a `CKR_OPERATION_NOT_INITIALIZED` during test of C_Digest.

Using:

- Debian 9 "Stretch" with in-kernel TPM resource manager (/dev/tpmrm0, kernel 4.19.0-0.bpo.5-amd64)
- opensc package version 0.16.0-3+deb9u1
- latest master of tpm2-tools (a76d317458f14cec0256b985987a6a2342d59a94)
- latest master of tpm2-pkcs11 (69bc971e020fc3bc839c31d4dbb1b6a66c1492d2)
- Infineon SLB9670 TPM built into a Lenovo P51 laptop (version information in terminal output below)

Terminal output:

```
$ TPM2_PKCS11_TCTI=device:/dev/tpmrm0 ./tools/tpm2_ptool.py init --pobj-pin=mypobjpin --owner-auth=ownerpass
Created a primary object of id: 1

$ TPM2_PKCS11_TCTI=device:/dev/tpmrm0 ./tools/tpm2_ptool.py addtoken --pid=1 --pobj-pin=mypobjpin --sopin=mysopin --userpin=myuserpin --label=labelauto-detecting TPM encryptdecrypt interface for wrapping key usage
Using "Software" based object authorization protections
Created token label: label

$ TPM2_PKCS11_TCTI=device:/dev/tpmrm0 pkcs11-tool --module /usr/local/lib/libtpm2_pkcs11.so --login --pin=myuserpin --keypairgen
Using slot 0 with a present token (0x1)
WARNING:esys:src/tss2-esys/api/Esys_CreateLoaded.c:359:Esys_CreateLoaded_Finish() Received TPM Error 
ERROR:esys:src/tss2-esys/api/Esys_CreateLoaded.c:129:Esys_CreateLoaded() Esys Finish ErrorCode (0x000b0143) 
ERROR: Esys_CreateLoaded: 0xb0143
ERROR: create_loaded 0xb0143
ERROR: Failed to generate key
error: PKCS11 function C_GenerateKeyPair failed: rv = CKR_GENERAL_ERROR (0x5)

Aborting.

$ TPM2_PKCS11_TCTI=device:/dev/tpmrm0 ./tools/tpm2_ptool.py addkey --algorithm=rsa2048 --label=label --userpin=myuserpin --key-label=label_asymmetric
Added key as label: "label_asymmetric"

$ TPM2_PKCS11_TCTI=device:/dev/tpmrm0 pkcs11-tool --module /usr/local/lib/libtpm2_pkcs11.so --pin=myuserpin --list-slots
Available slots:
Slot 0 (0x1): label                           Infineon
  token label        : label
  token manufacturer : Infineon
  token model        : SLB9670
  token flags        : rng, login required, PIN initialized, token initialized
  hardware version   : 1.16
  firmware version   : 7.61
  serial num         : 0000000000000000
error: PKCS11 function C_OpenSession failed: rv = CKR_SLOT_ID_INVALID (0x3)

Aborting.

$ TPM2_PKCS11_TCTI=device:/dev/tpmrm0 pkcs11-tool --module /usr/local/lib/libtpm2_pkcs11.so --login --pin=myuserpin --list-objects
Using slot 0 with a present token (0x1)
Private Key Object; RSA 
  label:      label_asymmetric
  ID:         36663435393064636532303231343132
warning: PKCS11 function C_GetAttributeValue(DECRYPT) failed: rv = CKR_ATTRIBUTE_TYPE_INVALID (0x12)

warning: PKCS11 function C_GetAttributeValue(SIGN) failed: rv = CKR_ATTRIBUTE_TYPE_INVALID (0x12)

warning: PKCS11 function C_GetAttributeValue(UNWRAP) failed: rv = CKR_ATTRIBUTE_TYPE_INVALID (0x12)

warning: PKCS11 function C_GetAttributeValue(DERIVE) failed: rv = CKR_ATTRIBUTE_TYPE_INVALID (0x12)

  Usage:      none
warning: PKCS11 function C_GetAttributeValue(ALWAYS_AUTHENTICATE) failed: rv = CKR_ATTRIBUTE_TYPE_INVALID (0x12)

Public Key Object; RSA 2048 bits
  label:      label_asymmetric
  ID:         36663435393064636532303231343132
warning: PKCS11 function C_GetAttributeValue(ENCRYPT) failed: rv = CKR_ATTRIBUTE_TYPE_INVALID (0x12)

warning: PKCS11 function C_GetAttributeValue(VERIFY) failed: rv = CKR_ATTRIBUTE_TYPE_INVALID (0x12)

warning: PKCS11 function C_GetAttributeValue(WRAP) failed: rv = CKR_ATTRIBUTE_TYPE_INVALID (0x12)

  Usage:      none

$ TPM2_PKCS11_TCTI=device:/dev/tpmrm0 pkcs11-tool --module /usr/local/lib/libtpm2_pkcs11.so --label="label" --test --pin myuserpin
Using slot 0 with a present token (0x1)
C_SeedRandom() and C_GenerateRandom():
  seems to be OK
Digests:
  all 4 digest functions seem to work
  SHA-1: OK
  ERR: C_Digest() didn't return CKR_OK for a NULL output buffer, but CKR_OPERATION_NOT_INITIALIZED (0x91)
  ERR: digest operation ended prematurely
Signatures (currently only RSA signatures)
warning: PKCS11 function C_GetAttributeValue(SIGN) failed: rv = CKR_ATTRIBUTE_TYPE_INVALID (0x12)

  testing key 0 (label_asymmetric)  -- can't be used for signature, skipping
Signatures: no private key found in this slot
Verify (currently only for RSA):
  testing key 0 (label_asymmetric)
warning: PKCS11 function C_GetAttributeValue(SIGN) failed: rv = CKR_ATTRIBUTE_TYPE_INVALID (0x12)

 -- can't be used to sign/verify, skipping
Unwrap: not implemented
Decryption (RSA)
warning: PKCS11 function C_GetAttributeValue(DECRYPT) failed: rv = CKR_ATTRIBUTE_TYPE_INVALID (0x12)

  testing key 0 (label_asymmetric)  -- can't be used to decrypt, skipping
2 errors
```

**Comment by PeterHuewe** [comment]
Hi,

the most basic problem is that the key generation fails.
The command fails with "0xb0143" which means "command code not supported".
The reason for this is, that for key generation it uses the 'TPM2_CreateLoaded' Command which was introduced with TPM Spec Revision 1.38.
Your TPM has spec revision 1.16.
Maybe you can find an firmware upgrade package at your Laptop Manufacturer support page.

Nevertheless, I'll look into whether we can workaround this somehow for TPMs with Spec Version 1.16.

**Comment by pfyvie** [comment]
Thanks for that clarification. The P51's TPM is already upgraded to its latest version, so we can assume it won't go past spec version 1.16 and the key generation via pkcs11-tool won't work for the foreseeable future. That's acceptable given that we have tpm2_ptool's addkey method, which does work (see above).

Even with that, we still see the other issues here - those being:

- After generating a key using `tpm2_ptool.py addkey` (which successfully uses the `tpm2_create` utility), that key is recognized as being on the virtual token by `--list-objects` but none of its capabilities can be enumerated (all enumeration calls throw `CKR_ATTRIBUTE_TYPE_INVALID (0x12)`) and the call concludes with `Usage:      none` for both public and private parts.
- Likely related to the failure to enumerate, self-test of signing capabilities fails against the generated RSA keys, throwing `can't be used for signature`/`to sign/verify`/`to decrypt, skipping`.
- Digest self-test ends prematurely. `tpm2_getcap -c algorithms` shows that the SLB9670 supports algorithms with hash capability sha1, hmac, keyedhash, xor, sha256, kdf1_sp800_56a, kdf1_sp800_108. Yet we see that after sha1 is tested OK, `C_Digest() didn't return CKR_OK for a NULL output buffer, but CKR_OPERATION_NOT_INITIALIZED (0x91)` is thrown.
- Enumeration of slots using `--list-slots` throws `CKR_SLOT_ID_INVALID (0x3)` after the slot information is output (not sure what it's trying to do here).

I'm not sure what the above symptoms indicate or how to begin troubleshooting them.

**Comment by williamcroberts** [comment]
Yep its:
```
rmt:error(2.0): command code not supported.
```
The workaround would be just to issue a create followed by a load command if this error is returned by the TPM or just test the firmware version on tpm code init.

**Comment by williamcroberts** [comment]
<snip>
> I'm not sure what the above symptoms indicate or how to begin troubleshooting them.

@pfyvie  this is going to be a code change to the create command invocation. @pdxjohnny is pulling bindings into the project, so it should be easier to address soon.

**Comment by williamcroberts** [comment]
I can also update the pkcs11 C code that if createloaded is not available to call create + load.

**Comment by williamcroberts** [comment]
I kind of lost track of this bug, I think this should be fixed for 1.0
